### DISCUSSION TITLE
How to unit test EditorContent

### DISCUSSION THREAD

**Discussion by jacobleetodd** [ueberdosis/tiptap#4008] [General Questions]
Hi There,

I'm using `EditorContent` and `useEditor` from `@tiptap/react` to build a basic rich text editor which is replacing a MUI `TextField` component. This results in the following default HTML rendered to the screen:
```
<div contenteditable="true" translate="no" class="ProseMirror" tabindex="0">
    <p>
        <br class="ProseMirror-trailingBreak">
    </p>
</div>
```

We use React Testing Library (RTL) and Jest for our unit tests. Since I am replacing a `TextField` there are a number of existing test suites that need to be updated. In order to do so, I need to select the appropriate DOM element and type something in it. I am able to select the div:
```
const editor = container.getElementsByClassName("ProseMirror")[0]
```
Or the p tag:
```
const ptag = container.querySelector('p')
```
But when I try to type in either of them:
```
await user.type(editor, "abc");
```
The tests fail and light up with errors:
```
  console.error
    Error: Uncaught [TypeError: (intermediate value)(intermediate value)(intermediate value).elementFromPoint is not a function]
    
  console.error
    Error: Uncaught [TypeError: target.getClientRects is not a function]
```
The strange thing is that when I `screen.debug()` I can see that the text is in the DOM:
```
      <div
          class="ProseMirror ProseMirror-focused"
          contenteditable="true"
          tabindex="0"
          translate="no"
        >
          <p>
            <br
              class="ProseMirror-trailingBreak"
            />
          </p>
          <p>
            abc
          </p>
      </div>
```
These feel like internal errors or are perhaps related to rendering the component in jsdom. I haven't been able to find any documentation or really any mention of how to unit test around the `EditorContent` component. This seems like a common use case so I am very confused. Any help is greatly appreciated!

**Comment by AnnyCaroline** [comment]
Hi, @jacobleetodd . I'm having the same issue.

Are you using `@testing-library/user-event`, right? AFAIK, `userEvent.type` only works with inputs and textareas: https://testing-library.com/docs/ecosystem-user-event/#typeelement-text-options

I tried to use `fireEvent.keyPress` instead. I received no errors, but my modifications are ignored.

**Reply by jacobleetodd** [reply]
Yes exactly `userEvent.type`, we do this:
```
const user = userEvent.setup();
```

**Comment by phfarolfi** [comment]
I'm having the same problem... even a simple change doesn't work.

```
const { container } = render(<RichText {...props} initialValue="foo" />)
const content = container.querySelector('.ProseMirror')
const text = content.querySelector('p')

fireEvent.keyPress(text, { key: 'a', keyCode: 'keyA' })

expect(text).toMatchInlineSnapshot(`
    <p>
        foo
    </p>
`)
```

**Comment by josokinas** [comment]
Came here for the same reason 🤔

**Comment by vojvodics** [comment]
What I've found is that there seems to be an issue with happy-dom/jsdom. Prosemirror is relying on some layout APIs, which aren't/can't be mocked in a non-browser environment. I tried mocking those APIs manually but didn't find a solution, so in the end, we switched to Cypress component tests for testing the editor.

**Comment by gdfreitasdev** [comment]
I was having the same issue and the snippet below met my needs. [Credits to **olpa** on StackOverflow](https://stackoverflow.com/a/77219899/5135113)

```ts
function getBoundingClientRect(): DOMRect {
  const rec = {
    x: 0,
    y: 0,
    bottom: 0,
    height: 0,
    left: 0,
    right: 0,
    top: 0,
    width: 0,
  };
  return { ...rec, toJSON: () => rec };
}

class FakeDOMRectList extends Array<DOMRect> implements DOMRectList {
  item(index: number): DOMRect | null {
    return this[index];
  }
}

document.elementFromPoint = (): null => null;
HTMLElement.prototype.getBoundingClientRect = getBoundingClientRect;
HTMLElement.prototype.getClientRects = (): DOMRectList =>
  new FakeDOMRectList();
Range.prototype.getBoundingClientRect = getBoundingClientRect;
Range.prototype.getClientRects = (): DOMRectList =>
  new FakeDOMRectList();
```

**Reply by kmmccafferty96** [reply]
This was working great until upgrading to a recent version (went from v2.1.13 to v2.4.0). Anyone have any ideas?

**Reply by sinnbeck** [reply]
This is my complete test that works for me in v2.4.0

```js
import React from 'react'
import {render, waitFor} from '@testing-library/react'
import {describe, expect, it, vi} from 'vitest'
import TiptapEditor from '@/Components/Editor/TiptapEditor.jsx'
import {userEvent} from '@testing-library/user-event'

// Mock the MenuBar component
vi.mock('@/Components/Editor/MenuBar', () => ({
    default: vi.fn(() => <div data-testid="menu-bar"></div>),
}))

function getBoundingClientRect() {
    const rec = {
        x: 0,
        y: 0,
        bottom: 0,
        height: 0,
        left: 0,
        right: 0,
        top: 0,
        width: 0,
    }
    return {...rec, toJSON: () => rec}
}

class FakeDOMRectList extends DOMRect {
    item(index) {
        return this[index]
    }
}

document.elementFromPoint = () => null
HTMLElement.prototype.getBoundingClientRect = getBoundingClientRect
HTMLElement.prototype.getClientRects = () => new FakeDOMRectList()
Range.prototype.getBoundingClientRect = getBoundingClientRect
Range.prototype.getClientRects = () => new FakeDOMRectList()

describe('TiptapEditor', () => {
    it('should render correctly with given content and mentions', async () => {
        const mockContent = '<p>Hello, world!</p>'
        const mockOnChange = vi.fn()
        const mockMentions = [{id: 1, name: 'John Doe'}]

        const {container, getByTestId} = render(
            <TiptapEditor
                content={mockContent}
                onChange={mockOnChange}
                mentions={mockMentions}
            />
        )

        // Check if the MenuBar is rendered
        expect(getByTestId('menu-bar')).toBeInTheDocument()

        // Check if the EditorContent is rendered
        const editor = await waitFor(() =>
            container.querySelector('.ProseMirror').querySelector('p')
        )

        expect(editor).toBeInTheDocument()
    })

    it('should call onChange with the updated HTML content', async () => {
        const mockContent = ''
        const mockOnChange = vi.fn()
        const mockMentions = [{id: 1, name: 'John Doe'}]

        const {container} = render(
            <TiptapEditor
                content={mockContent}
                onChange={mockOnChange}
                mentions={mockMentions}
            />
        )
        const user = userEvent.setup()

        const editor = await waitFor(() =>
            container.querySelector('.ProseMirror').querySelector('p')
        )

        await user.type(editor, 'New content')

        expect(editor).toMatchInlineSnapshot(`
    <p>
      New content
    </p>
`)

        expect(mockOnChange).toHaveBeenCalledTimes(11)
        expect(mockOnChange).toHaveBeenCalledWith('<p>New content</p>')
    })
})

**Comment by sross0988** [comment]
> I was having the same issue and the snippet below met my needs. [Credits to **olpa** on StackOverflow](https://stackoverflow.com/a/77219899/5135113)
> 
> ```ts
> function getBoundingClientRect(): DOMRect {
>   const rec = {
>     x: 0,
>     y: 0,
>     bottom: 0,
>     height: 0,
>     left: 0,
>     right: 0,
>     top: 0,
>     width: 0,
>   };
>   return { ...rec, toJSON: () => rec };
> }
> 
> class FakeDOMRectList extends Array<DOMRect> implements DOMRectList {
>   item(index: number): DOMRect | null {
>     return this[index];
>   }
> }
> 
> document.elementFromPoint = (): null => null;
> HTMLElement.prototype.getBoundingClientRect = getBoundingClientRect;
> HTMLElement.prototype.getClientRects = (): DOMRectList =>
>   new FakeDOMRectList();
> Range.prototype.getBoundingClientRect = getBoundingClientRect;
> Range.prototype.getClientRects = (): DOMRectList =>
>   new FakeDOMRectList();
> ```

For those wanting more context on how to use this with `@testing-library/react`. You can add this code to your test file to remove any rendering errors and then test changing the editor component with:

```
      const emailEditor = container.querySelector('.ProseMirror').querySelector('p');

      fireEvent.change(emailEditor, {
        target: { textContent: 'To whom it may concern...' }
      });
      
```

It took me a lot of searching and testing different combinations to get this right so I hope this helps someone. 

Thanks to everyone else for all the useful info on this thread!

**Comment by McKayMower** [comment]
if you are using useEditor, you can do it the following way:

```
interface Props {
  testid: string
  // other props
}

export const Tiptap = ({testid, ...props}: Props) => {
   const editor = useEditor({
    // other editor config
    editorProps: {
      attributes: {
        testid: testid
      },
    },
  });
  
  return (
     <EditorContent
        editor={editor}
      />
  )
}
```

then, you can go inspect the HTML in localhost or in your test environment to validate the testid appears as an attribue. hope this helps!

**Comment by tonai** [comment]
I did have the same issue and I added the proposed code by @gdfreitas into a jest setup file and the test does not fail anymore.
But I noticed there a strange behaviour because there an empty paragraph before what you type.
Here is my test file:
```tsx
import { describe, test } from "@jest/globals";
import { render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { EditorProvider } from "@tiptap/react";
import StarterKit from "@tiptap/starter-kit";

describe("RichTextEditor component", () => {
  test("it returns empty paragraph", async () => {
    const user = userEvent.setup();
    render(
      <EditorProvider
        extensions={[StarterKit.configure()]}
        content=""
      ></EditorProvider>,
    );
    await waitFor(() => expect(screen.queryByRole("textbox")).toBeDefined());
    const content = screen.getByRole("textbox");
    console.log(content.innerHTML); // <p><br class="ProseMirror-trailingBreak"></p>
    await user.type(content, "a");
    console.log(content.innerHTML); // <p><br class="ProseMirror-trailingBreak"></p><p>a</p>
  });
});

```

* The first log is `<p><br class="ProseMirror-trailingBreak"></p>` which is fine.
* The second log is `<p><br class="ProseMirror-trailingBreak"></p><p>a</p>` but I didn't press any enter key so I should only have got `<p>a</p>`.

**Reply by MirkoMContinuum** [reply]
I'm having the same issue, did you find a solution?

**Comment by matheusramos** [comment]
I've implemented a controlled editor and my workaround for testing is setting the `role` and `aria-label` manually:

```ts
 const handleTransaction = ({ editor }: EditorEvents['transaction']) => {
    if (editor.isEmpty) {
      onChange?.('');
      return;
    }
    onChange?.(editor.getHTML());
  };

  const editor = useEditor({
    extensions: [
      StarterKit,
    ],
    content: value,
    onTransaction: handleTransaction,
    editorProps: {
      attributes: {
        'aria-label': ariaLabel,
        role: 'textbox',
      },
    },
  });

  useEffect(() => {
    if (!editor) {
      return;
    }
    editor.commands.setContent(value || '', false, { preserveWhitespace: 'full' });
  }, [editor, value]);

  // ...
  return <EditorContent editor={editor} />;
```

Then test it with something like
```tsx
  it('should update content when typing', async () => {
    render(<TipTapEditor value="" onChange={onChange} aria-label="Editor" />);

    const editor = screen.getByRole('textbox', { name: 'Editor' });
    await userEvent.type(editor, 'Hello world');

    expect(onChange).toHaveBeenCalledWith('<p>Hello world</p>');
    // or
    expect(editor).toHaveTextContent(/Hello world/);
  });
```

**Reply by MazenRefaat** [reply]
@matheusramos thanks for sharing 🙏 
QQ what version are you using? also aren't you seeing any errors when you use `userEvent.type` ?

**Comment by maciejkrol18** [comment]
I managed to get rid of all these errors using the mocks below (I'm using vitest though and I put them in `vitest.setup.ts`).
```tsx
Range.prototype.getBoundingClientRect = () => ({
  bottom: 0,
  height: 0,
  left: 0,
  right: 0,
  top: 0,
  width: 0,
  x: 0,
  y: 0,
  toJSON: vi.fn(),
});

Range.prototype.getClientRects = () => ({
  item: () => null,
  length: 0,
  [Symbol.iterator]: vi.fn(),
});

Document.prototype.elementFromPoint = vi.fn();
```

With this in place, I also set `role` of `textbox` to the prosemirror root element using `editorProps` property within the `useEditor` hook

```tsx
editorProps: {
  attributes: {
    role: "textbox",
  },
}
```

And then just use `user.type()` from `@testing-library/user-event` as follows:

```ts
user.type(screen.getByRole("textbox"), "Lorem ipsum")
```

No errors reported 👍


<details><summary>Versions of relevant packages</summary>

```json
"devDependencies": {
  "@testing-library/dom": "^10.4.0",
  "@testing-library/jest-dom": "^6.6.3",
  "@testing-library/react": "^16.3.0",
  "@testing-library/user-event": "^14.6.1",
  "@types/react-dom": "^19.0.4",
  "@vitejs/plugin-react": "^4.7.0",
  "jsdom": "^26.1.0",
  "msw": "^2.10.4",
  "vite-tsconfig-paths": "^5.1.4",
  "vitest": "^3.2.4"
},
```

</details>
